ApiKey owner (org) scope UI work

chenriksson opened the pull request:
Issue #4872.

Work remaining for API key owner (org) scopes:
- Auditing (#4958)
- Remove deprecated Scope.cs ctor (will do in separate PR, to minimize this PR)
- Integrate with PermissionsService

chenriksson commented:
- Create key has same UI if user is not member of any organization:

![image](https://user-images.githubusercontent.com/15364949/32473245-7f45dd54-c31b-11e7-811c-329fbb38118c.png)

- Existing v2 key w/o owner scope:

![image](https://user-images.githubusercontent.com/15364949/32473080-995b41ee-c31a-11e7-8341-fa819e0d6bf3.png)

- Existing v2 key with owner scope:

![image](https://user-images.githubusercontent.com/15364949/32473103-ac3fee40-c31a-11e7-838d-180edb99b79a.png)

- Creation of new v2 key, will always have owner scope:

![image](https://user-images.githubusercontent.com/15364949/32473154-f51187be-c31a-11e7-8f72-5c1e679106d4.png)

- Other Notes:
  - Push new action disabled for organization collaborators
  - Packages displayed depends on the selected owner

scottbommarito commented on src/NuGetGallery/Controllers/UsersController.cs:
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
         [HttpPost]
         [ValidateAntiForgeryToken]
-        public virtual async Task<JsonResult> GenerateApiKey(string description, string[] scopes = null, string[] subjects = null, int? expirationInDays = null)
+        public virtual async Task<JsonResult> GenerateApiKey(string description, string owner, string[] scopes = null, string[] subjects = null, int? expirationInDays = null)
```
I think we should change `scopes` to be `actions` here and elsewhere. It's incredibly confusing that a `Scope` is an object that contains an owner, a set of actions, and a list of subjects AND a "scope" is an action that the API key allows on the package. Now we're also referring to the owner as a "scope" too.

chenriksson replied:
I agree. The term "scope" is used interchangeably with scope action throughout the code.

I'm ok with fixing, but prefer to do it in a separate PR as I don't want to inflate this one.

scottbommarito commented on src/NuGetGallery/Scripts/gallery/page-api-keys.js:
```diff
@@ -106,13 +101,33 @@
+            }, this);
+
+            // Do not default to push new scope if disabled (organization collaborator)
+            this.PushNewEnabled = ko.pureComputed(function () {
```
You will probably want to do this for all three possible actions (push new, push version, and unlist) because the `PermissionsService` allows for it.

chenriksson replied:
Push version and unlist are always available, even to organization collaborators. I only need to determine when to enable/disable push new.

scottbommarito replied:
always available NOW

in the future they may not be, so we use `PermissionsService` to centralize this logic

scottbommarito commented on src/NuGetGallery/Views/Users/ApiKeys.cshtml:
```diff
@@ -174,6 +174,10 @@
                     </li>
                 </ul>
+                <!-- ko if: Owner -->
+                <b>Package owner:</b> <span data-bind="text: Owner"></span>
```
any thoughts on showing the current user if there is no owner scope?

chenriksson replied:
We decided that no owner scope means that it can match any, not just the current user. That's why I chose to show "*"... but am open to a different suggestion.

scottbommarito replied:
Oh my bad, I will need to update my PR to understand this logic

scottbommarito replied:
according to @anangaur, we decided that no owner scope means only match current user

personally, I prefer this because it forces users to use organizations correctly

scottbommarito commented on src/NuGetGallery/Controllers/UsersController.cs:
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
             }
 
+            // Get the owner scope
+            User scopeOwner = string.IsNullOrEmpty(owner)
```
We need to check that the current user actually has the permissions to create an API key with this user as their owner scope. From what I can see here, someone could create an API key with any user as the owner in the scope.

On a similar note, we should check that the user has permissions to create an API key with these scopes (actions) on the owner (e.g. organization collaboration shouldn't have push new).

joelverhagen replied:
Yes, client side validation is a "hint" to the user. Server side validation is the real enforcement.

chenriksson replied:
Adding server side validation, but will integrate w/ PermissionsService in follow-up PR since it's under flux right now.

scottbommarito commented on src/NuGetGallery/Views/Users/ApiKeys.cshtml:
```diff
@@ -293,6 +297,23 @@
             }
         </div>
+
+        @if (Model.PackageOwners.Count > 1)
```
If `Model.PackageOwners.Count == 1` you may want to pass the single option as a hidden item on the form.

It's not needed because you have the `GetCurrentUser` call in `GenerateApiKey`, but in case some code changes make it possible for the only package owner to be NOT the current user, this is more consistent.

scottbommarito replied:
https://github.com/NuGet/NuGetGallery/pull/4963/files#diff-5058b8e804fd7b6e9e60a29de7f6fff6R60 is how I did it for upload/edit

chenriksson replied:
Modified to still populate PackageOwner observable when single owner (self), but will hide the drop-down.

scottbommarito commented on src/NuGetGallery/ViewModels/ScopeViewModel.cs:
```diff
@@ -5,11 +5,13 @@ namespace NuGetGallery
-        public ScopeViewModel(string subject, string allowedAction)
+        public ScopeViewModel(string owner, string subject, string allowedAction)
         {
+            Owner = owner ?? "*";
```
`"*"` -> `NuGetPackagePattern.AllInclusivePattern`

chenriksson replied:
It's not a package pattern, but will add AllInclusivePattern const to the view model.

chenriksson replied:
Removed per @anangaur feedback that legacy keys match only current user.

joelverhagen commented on src/NuGetGallery/Scripts/gallery/page-api-keys.js:
```diff
@@ -106,13 +101,33 @@
+                return self.PackageOwner() && self.PackageOwner().CanPushNew;
+            }, this);
+            this.PushNewEnabled.subscribe(function (newValue) {
+                var defaultPushId = (newValue)
```
nit: why parens around `newValue`?

chenriksson replied:
Updated

joelverhagen commented on src/NuGetGallery/Controllers/UsersController.cs:
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
+            User scopeOwner = string.IsNullOrEmpty(owner)
+                ? GetCurrentUser()
+                : _userService.FindByUsername(owner);
+            if (scopeOwner == null)
```
Just making sure -- this code is saying all future API keys must be scoped to a single owner?

chenriksson replied:
Yes, API keys will be scoped to a single owner.

joelverhagen commented:
We need a final sign-off from all feature owners and @anangaur that:

1. What do API keys without an owner scope do when pushing a new ID and the user is part of at least one org. _This is the ambiguous case..._
1. What should the API key UI show for the scopes/owners/actions for these old owner-less API keys?

I keep forgetting our conclusion on these discussions 😊.

scottbommarito commented:
@joelverhagen 
question 1 is handled in https://github.com/NuGet/NuGetGallery/pull/4963 (we should ask it there)
not sure about question 2

anangaur commented:
> What do API keys without an owner scope do when pushing a new ID and the user is part of at least one org. This is the ambiguous case...
What should the API key UI show for the scopes/owners/actions for these old owner-less API keys?

Both cases are for existing API keys and they should get associated to the account owner's account, UI or otherwise. The spec has this discussed here: https://github.com/NuGet/Home/wiki/Organizations-on-NuGet.org#what-happens-to-authors-existing-api-keys

scottbommarito commented on src/NuGetGallery/Scripts/gallery/page-api-keys.js:
```diff
@@ -139,7 +154,7 @@
                 }, self);
             }
-
+            
```
nit: unnecessary whitespace

chenriksson replied:
reverted

chenriksson commented on src/NuGetGallery/Controllers/UsersController.cs:
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
+                : _userService.FindByUsername(owner);
+            if (scopeOwner == null)
+            {
+                Response.StatusCode = (int)HttpStatusCode.NotFound;
```
Updated

chenriksson commented on src/NuGetGallery/Controllers/UsersController.cs:
```diff
@@ -105,23 +105,33 @@ public virtual ActionResult ApiKeys()
-                .OrderBy(i => i)
+            // Get package owners (user's self or organizations)
+            var owners = user.Organizations
+                .Select(o => CreateApiKeyOwnerViewModel(o.Organization, canPushNew: o.IsAdmin))
```
Will integrate w/ PermissionsService in a follow-up PR.

chenriksson commented:
@scottbommarito @joelverhagen Addressed feedback. I will integrate it with PermissionsService in a separate PR, to avoid taking dependency for now on #4963.

There's still a test I'd like to add for server-side validation, but I'm currently blocked there by a VS bug with test discovery.

scottbommarito commented on src/NuGetGallery/Authentication/AuthenticationService.cs:
```diff
@@ -524,7 +527,11 @@ public virtual CredentialViewModel DescribeCredential(Credential credential)
                 Value = kind == CredentialKind.Token && credential.Description == null ? credential.Value : null,
-                Scopes = credential.Scopes.Select(s => new ScopeViewModel(s.Subject, NuGetScopes.Describe(s.AllowedAction))).ToList(),
+                Scopes = credential.Scopes.Select(s => new ScopeViewModel(
+                        scopeOwner.Username,
```
You shouldn't cache the value and use it for each scope in the list when you're enumerating through each scope anyway.

```
s.Owner?.Username ?? credential.User
```

scottbommarito commented on src/NuGetGallery/Extensions/ScopeExtensions.cs:
```diff
@@ -70,6 +71,19 @@ public static bool HasOwnerScope(this Scope scope)
             return scope.OwnerKey.HasValue;
         }
+
+        public static User GetOwnerScope(this IEnumerable<Scope> scopes)
```
I would suggest removing this function and using `scopes.SingleOrDefault(s => s.Owner)` instead in all the places that you use it for the following reasons.

- it can be reduced into the one-liner above
- it will need to be removed when we allow multiple owners per scope
- its existence makes it very easy to not treat scopes appropriately (e.g. by assuming that the first owner in the scope applies to ALL scopes)

chenriksson replied:
IMHO, if we eventually allow multiple scope owners, having this in a central location would actually help identify callers that need updating.

scottbommarito commented on src/NuGetGallery/Strings.resx:
```diff
@@ -618,4 +618,7 @@ For more information, please contact '{2}'.</value>
     <value>The account:{0} was deleted succesfully.</value>
   </data>
+  <data name="ApiKeyOwnerRequired" xml:space="preserve">
+    <value>Can't generate an API key without a package owner.</value>
```
`You must specify a package owner to generate an API key.` might be better

chenriksson replied:
Was consistent with ApiKeyDescriptionRequired, but I have no strong preference. Updated.

joelverhagen commented on src/NuGetGallery/ViewModels/ApiKeyViewModel.cs:
```diff
@@ -20,6 +20,13 @@ public ApiKeyViewModel(CredentialViewModel cred)
             }
 
+            // Currently ApiKeys.cshtml has single Owner per ApiKey restriction.
+            var owner = cred
```
Can you use `GetOwnerScope`  extension method here?

chenriksson replied:
No... code looks the same, but this is querying on view models not entities.
